README
QFE - Qt FTN Message Editor (C) 2005-2008 by Alexander Shiyan
--------------------------------------------------------------------------------

[ *** Description *** ]
 QFE is an FTN message editor. A full-featured cross-platform FTN message
editor with a graphical interface. It written on C++/Qt and does not depend on
either KDE, Gnome or other window managers.
 Fido-specific parameters readed by fidoconfig package from Husky project.
I write qfe for Linux, but it works also on Win32 and MacOS X, and in theory,
it should compile & work on all other Qt-supported platforms.
 To learn more about 'QFE' please see the various documents in the doc folder.

[ *** Requirements *** ]
 QFE is based on Trolltech's Qt library 3.2.x or above, but 4.x.x is not
supported yet, so be sure you have the correct library version installed.
To compile QFE from sources You also need have installed Qt devel headers.
 QFE now sucessfully compiled with:
 On Linux:   GNU GCC   (3.x.x & 4.x.x),
             Intel C++ (9.x).
 On MacOS X: GNU GCC (3.3.x).
 On Windows: MinGW GCC (3.3.x).
 Also You need fidoconfig package from husky project (http://husky.sourceforge.net/).

[ *** License *** ]
 QFE is released under GNU General Public License (GPL).
For more information about GPL, see http://www.gnu.org/licenses/gpl.html

[ *** Quick-Compile & Install Guide *** ]
 tar -xf qfe-0.5.1.tar.bz2
 cd qfe
 ./configure
 make
 make install #As a user with proper file system permissions

 For link fidoconfig library statically, you must add parameter "staticlibs" to
qmake, e.g. qmake "CONFIG+=debug_off staticlibs"

 QFE must be compiled with lot of debug messages, printed to stdout. Debug
switches on if "-d" parameter passed to commandline, e.g. ./qfe -d

 By default, program installs in /usr/local. For install in another directory
You must (before doing qmake) export QFE_PREFIX variable with base install path,
e.g. export QFE_PREFIX=/usr, or use commandline option '--prefix' for configure
script.
 Some another installation tips located in FAQ. Read it first.

[ *** Help Wanted *** ]
 There are many ways to support this program:
* Creating packages for the various distributions.
* Translate 'QFE' to your language.
* Testing, searching for bugs and sending bug reports.
* But also every idea (or even patch) to improve the program is always welcome!

[ *** Bugs *** ]
 QFE is a new project, so you may find many problems. If you can't find the
solution, please contact us.
 Please also report bugs and/or wishes with the 'QFE' bug tracker
at http://sourceforge.net/tracker/?group_id=132811 or mail me at
shc@users.sourceforge.net or at 2:5030/39.2 by fidonet.
My ICQ UIN for contacts is 30092155.
 Patches for QFE are even more appreciated.

--------------------------------------------------------------------------------
http://qfe.sourceforge.net/


ChangeLog
QFE version SVN ==============================================================
...

QFE version 0.5.2 === Released 20.02.2008 ====================================
- Fix decode "FMPT" and "TOPT" kludges in MSG-areas.
- Fixes to compile on FreeBSD.
- Fix crash on empty included file in fidoconfig.
- Patch for HUSKY no more be needed.

QFE version 0.5.1 === Released 27.01.2007 ====================================
- Fix update squish base UID when write new message.
- Fix write new message & delete message in Jam-areas.
- Fix UU & Base64 decoder for work with badly-encoded parts.
- Rewrite "CHRS" and "CHARSET" kludge handler.
- Other minor fixes.

QFE version 0.5 === Released 30.05.2006 ======================================
- Fix crash when append address from main list into AddressBook.
- Added support for view JPEG-images in GIF-kludge.
- Added example fidoconf configuration file.
- Added "group description" field to area-property dialog.
- Added Geek viewer (Read about 'Geek' on http://www.geekcode.com/).
- Added "Summary base size" field in area property dialog.
- Win32 binary version now contain some HUSKY-related utilites.
- Implement "Archive area". Important messages may stored in this local area.
- Proper error messages when errors found in fidoconf configuration file.
- Added "clipboard.tpl" template for format text inserted from clipboard.
- Extend some debug messages in "Debug"-mode.
- Fixes for PID-kludge as described in FSC-0046.
- Fixes for quoting as described in FSC-0032.
- QFE now not use SMAPI! We use own classes and functions for access to
  messagebases.
- Added "-l" command line option, which override language for interface.
- Fixes for compile on MacOS X.
- Fixes for compile on x86_64 platform.
- Update Binreloc library.
- Major fixes and enhancements.

QFE version 0.4.6 === Released 12.12.2005 ====================================
- Fixed incorrect handle "Read/Unread" flags.
- Added Nodelist browser.
- Added "-h" (Help) command line option.
- Added "@hostname@" macro.
- Added "-e" switch to commandline options, allow jump to selected area after
  program started.
- First Win32 public binary release.

QFE version 0.4.5 === Released 31.10.2005 ====================================
- Keep & update "RFC-References"-kludge.
- Added summary statistics to area-property dialog.
- Added option for view kludges in editor on reply. (Feature request #1219855)
- Fixed decode more than one UUE-sections in single message.
- Fixed quoted text wrapping.
- Speedup areas rescan.
- Added "Add sender to AddressBook" option to popupmenu on selected message.
- Added progress indicator to bottom bar (for rescan & find).
- Added option for replace smilies to icons in viewer window.
- Major fixes and enhancements.

QFE version 0.4.3 === Released 27.09.2005 ====================================
- Added first-start welcome dialog.
- Added support for HTML-messages.
- Block running second instance of QFE (All platforms).
- Fixes for "Intel C++" and "Microsoft Visual Studio 2003 .NET" compilers.
- Fixed store/restore window position and size.
- Debug information now puts only to file (~/qfe/debug.log).
- Minor fixes.

QFE version 0.4.2 === Released 01.08.2005 ====================================
- Fixed wrong regular expression for check email-address in addressbook.
- Multimedia bitmaps & sounds support, such as in FIPS, DVMEDIT.
- Handle messages with "RRQ".
- Added "Confirmations" page in options window.
- EMail addresses now clickable in view vindow. Configuration option "External
  EMail program" added to Options dialog.
- Minor fixes.

QFE version 0.4.1 === Released 11.07.2005 ====================================
- Fixed view messages with alien charsets in subject line.
- Fixed rescanning when window is taken from focus.
- Fixed wrong MSGID generation.
- Added option "Next unread message" to main vindow. (Feature request #1218490)
- Fixed incorrect Url parsing. (Bug #1218473)
- After wrote message, rescan only modified areas. (Feature request #1217322)
- Fixed incorrect wordwrap in viewer. (Bug #1217323)
- Fixed incorrect template update.

QFE version 0.4 === Released 08.06.2005 ======================================
- Russian translation corrected.
- Added "--configdir" startup option.
- Handle "kludges.lst" for append individual kludges in message.
- Block running second instance of QFE (Linux & Mac only).
- Implemented random Origin insertion in message editor.
- "Print message" option completed.
- "Save message" option completed. Supported formats are TXT, HTML.
- Added "Templates" page in options window. It contains template editor.
- Hyperlinks now clickable in view vindow. Configuration option "External HTML
  browser" added to Options dialog.
- Minor fixes and enhancements. Speedup rescan areas.

QFE version 0.3 === Released 18.05.2005 ======================================
- Added "@command@" macro.
- Add first-time start "select config" dialog.
- Check invalid characters in output name for UU/Base64-decoded files.
- QFE now compiles & works on Win32.
- Minor fixes.

QFE version 0.2 === Released 22.04.2005 ======================================
- Handle EchoTossLogFile (fidoconfig).
- Added "Insert text file" option to editor window.
- Added addressbook.
- Added "Find message" dialog.
- Added highlight in MessageList if mail is for us.
- Added highlight URLs in view windows.
- Hide "Via" kludges.
- QFE now may link static fidoconfig-library.
- Added "Wrap Text" option in view window.
- Added "Unsubscribe" option to AreaList.
- Added UU && Base64 decoder.
- Fixed incorrect view multiline "from", "to", "subject" etc..
- Install fixes.
- Major enhancements and bug fixes.

QFE version 0.1 === Released 30.03.2005 ======================================
- First public version.

QFE develop started 03.03.2005 ===============================================


src/tconfig.h
/***************************************************************************
 *   Copyright (C) 2005-2008 by Alexander Shiyan                           *
 *   shc@users.sourceforge.net                                             *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#ifndef _TCONFIG_H_
#define _TCONFIG_H_

#include <qdir.h>
#include <qaction.h>
#include <qobject.h>
#include <qtoolbar.h>
#include <qlistview.h>
#include <qsplitter.h>
#include <qpopupmenu.h>
#include <qtextcodec.h>
#include <qstylesheet.h>
#include <qpixmapcache.h>
#include <qfontdatabase.h>
#include <qdesktopwidget.h>

#include "funchider.h"

class TConfig;

#include "utils.h"

#include "tarea.h"
#include "tbook.h"
#include "tfidoconfig.h"
#include "tinisettings.h"

#define CONFIG_SECTION_MAIN				"General"
#define CONFIG_SECTION_CONF				"Config"
#define CONFIG_SECTION_FIDO				"Fido"
#define CONFIG_SECTION_EDIT				"Editor"
#define CONFIG_SECTION_STYL				"StyleDlg"
#define CONFIG_SECTION_KEYB				"Keyboard"
#define CONFIG_SECTION_CONFIRM				"Confirm"

#define CONFIG_FIRSTRUN					"FirstRun"

#define CONFIG_WINDOWSTATE				"WindowState"

#define CONFIG_VIEWSTATUSBAR				"ViewStatusBar"
#define CONFIG_VIEWKLUDGES				"ViewKludges"
#define CONFIG_VIEWHIGHLIGHT				"ViewHighlight"
#define CONFIG_VIEWWORDWRAP				"ViewWordWrap"
#define CONFIG_VIEWORIGIN				"ViewOrigin"

#define CONFIG_FIDOCODEPAGE				"DefCodePage"
#define CONFIG_FIDODEFZONE				"DefZone"
#define CONFIG_FIDOCONFNAME				"ConfigName"

#define CONFIG_EXTBROWSER_CMD				"HtmlBrowserCommand"
#define CONFIG_EXTBROWSER_PRM				"HtmlBrowserParams"
#define CONFIG_EXTEMAIL_CMD				"EMailClientCommand"
#define CONFIG_EXTEMAIL_PRM				"EMailClientParams"

#define CONFIG_FONTNAME					"FontName"
#define CONFIG_FONTSIZE					"FontSize"
#define CONFIG_FONTCOLOR				"FontColor"
#define CONFIG_FONTBACK					"BackColor"
#define CONFIG_VIEWSTYLE				"HighlightStyle"

#define CONFIG_SEARCHAREA				"SearchArea"
#define CONFIG_SEARCHTEXT				"SearchText"
#define CONFIG_SEARCHCASE				"SearchCase"
#define CONFIG_SEARCHWHERE				"SearchWhere"
#define CONFIG_SEARCHREGEXP				"SearchRegExp"

#define CONFIG_REPLACELIST				"Replace"

#define CONFIG_SHOWKLUDGES				"ShowKldgesInEditor"

#define CONFIG_SMILIES_REPLACE				"ReplaceSmilies"

#define CONFIG_LASTAREA					"LastArea"
#define CONFIG_LASTOPENSAVE				"LastDir"

#define CONFIG_MEDIA_DIR				"MediaDir"
#define CONFIG_MEDIA_IMAGES				"MediaBmp"
#define CONFIG_MEDIA_WAVES				"MediaWav"

#define CONFIG_CONFIRM_DELETE				"DeleteMessage"
#define CONFIG_CONFIRM_QUIT				"Quit"
#define CONFIG_CONFIRM_QUITSAVE				"QuitSave"
#define CONFIG_CONFIRM_RRQ				"RRQ"

#define STYLE_BOLD					(1<<0)
#define STYLE_ITALIC					(1<<1)
#define STYLE_UNDERLINE					(1<<2)
#define STYLE_INVERSE					(1<<3)

#define MSG_OPERATION_NEW				(1<<0)
#define MSG_OPERATION_EDIT				(1<<1)
#define MSG_OPERATION_REPLY				(1<<2)
#define MSG_OPERATION_REPLYNET				(1<<3)
#define MSG_OPERATION_FORWARD				(1<<4)
#define MSG_OPERATION_UNSUBSCRIBE			(1<<5)
#define MSG_OPERATION_CONFIRM				(1<<6)

#define SYSOPIMAGE_WIDTH				110
#define SYSOPIMAGE_HEIGHT				110

#define TPL_MESSAGE_TPL					"message.tpl"
#define TPL_REPLY_TPL					"reply.tpl"
#define TPL_FORWARD_TPL					"forward.tpl"
#define TPL_UNSUBSCRIBE_TPL				"unsubscribe.tpl"
#define TPL_CONFIRM_TPL					"confirm.tpl"

#define TPL_ARCHIVE_TPL					"archive.tpl"

#define TPL_SAVETXT_TPL					"save_txt.tpl"
#define TPL_SAVEHTM_TPL					"save_htm.tpl"

#define TPL_CLIPBOARD_TPL				"clipboard.tpl"

#define TPL_ORIGIN_LST					"origin.lst"
#define TPL_KLUDGES_LST					"kludges.lst"

#define PROGRESS_TYPE_RESCAN				1
#define PROGRESS_TYPE_SEARCH				2
#define PROGRESS_TYPE_COMPILE				3
#define PROGRESS_TYPE_SAVENODES				4
#define PROGRESS_TYPE_LOADNODES				5

#define ADDRESSBOOK_NAME				"addressbook.xml"

#define DEBUG_LOG_FILE					"debug.log"

#define COMPILED_NODELIST_FILE				"nodelist.dat"

#define REGEXP_ADDRESS					"^[\\d]{1,4}:[\\d]{1,4}/[\\d]{1,4}(\\.[\\d]{1,4}|\\b)(@[.\\S\\w]{1,15}|\\b)"
#define REGEXP_FROMTO					"^[\\S].{1,36}$"
#define REGEXP_EMAIL					"\\b([-_./\\d\\w]|[\\d][:]){1,20}[@][-_\\d\\w]{1,10}(\\.[-_\\d\\w]{2,12}){1,5}"
#define REGEXP_URL					"((ftp|http|https)://|www\\.)([\\d\\S\\w]{2,}\\.)+[-/?&%_+=;:~a-zA-Z0-9]{2,}($|\\b)"

#define REGEXP_FIPS_MM					"FmM[BW](2|4|2[\\x20]).?[\\d]{7,7}"

#define REGEXP_UNICODE_CHRS				"&#[\\d]{2,5};"

#define MAIN_URL					"http://qfe.sourceforge.net/"
#define AUTHOR_EMAIL					"shc@users.sourceforge.net"
#define AUTHOR_FTN					"2:5030/39.2"
#define BUG_REPORT_URL					"http://sourceforge.net/tracker/?group_id=132811"

#define ERROR_INITFIDO_OK				0
#define ERROR_INITFIDO_CONFIG				1
#define ERROR_INITFIDO_LOCK				2

#if defined(Q_OS_MACX)
#	define PLATFORM_NAME				"MacOS"
#	define SHORT_PLATFORM_NAME			"Mac"
#	define DEFAULT_VIEW_FONT			"Monaco"
#	define DEFAULT_VIEW_FONTSIZE			12
#	define DEFAULT_EXT_BROWSER			"open -a Safari"
#	define DEFAULT_EXT_BROWSER_PRM			""
#	define DEFAULT_EXT_EMAIL			"open -a Mail"
#	define DEFAULT_EXT_EMAIL_PRM			""
#elif defined (Q_OS_WIN)
#	define PLATFORM_NAME				"Windows"
#	define SHORT_PLATFORM_NAME			"Win"
#	define DEFAULT_VIEW_FONT			"Lucida Console"
#	define DEFAULT_VIEW_FONTSIZE			10
#	define DEFAULT_EXT_BROWSER			"explorer.exe"
#	define DEFAULT_EXT_BROWSER_PRM			"%u"
#	define DEFAULT_EXT_EMAIL			"msimn.exe"
#	define DEFAULT_EXT_EMAIL_PRM			"/mailurl:mailto:%t?Subject=%s&Body=%B"
#elif defined(Q_OS_UNIX)
#	if defined(__x86_64__)
#		if defined(__FreeBSD__)
#			define PLATFORM_NAME		"Linux(64Bit)"
#			define SHORT_PLATFORM_NAME	"Lnx64"
#		else
#			define PLATFORM_NAME		"FreeBSD(64Bit)"
#			define SHORT_PLATFORM_NAME	"BSD64"
#		endif
#	else
#		if defined(__FreeBSD__)
#			define PLATFORM_NAME		"FreeBSD"
#			define SHORT_PLATFORM_NAME	"BSD"
#		else
#			define PLATFORM_NAME		"Linux"
#			define SHORT_PLATFORM_NAME	"Lnx"
#		endif
#	endif
#	define DEFAULT_VIEW_FONT			"Console"
#	define DEFAULT_VIEW_FONTSIZE			12
#	define DEFAULT_EXT_BROWSER			"firefox"
#	define DEFAULT_EXT_BROWSER_PRM			"%u"
#	define DEFAULT_EXT_EMAIL			"kmail"
#	define DEFAULT_EXT_EMAIL_PRM			"-s %s --body %B %t"
#else
#	error "Unable to determine target operating system!"
#endif

#if defined(Q_OS_WIN)
#	define FIXED_FILE_CASE(x)			x.lower()
#else
#	define FIXED_FILE_CASE(x)			x
#endif

#define QFE_LOGO_NAME					"qfe_logo.xpm"
#define QFE_LOGO					QPixmap::fromMimeSource(QFE_LOGO_NAME)

#define GET_CURRENT_LOCALE()				QTextCodec::locale()

typedef struct _TProgressCallBack_ {
	void *InitProgress(uint32_t);
	void *ProcProgress(uint32_t);
	void *DoneProgress();
} TProgressCallBack;

class TConfig
{
public:
	TConfig(QDir*, QDir*, QString = QString::null);
	~TConfig();

	ps_fidoconfig FidoConfig;

	QString ConfigPath;
	QString HomePath;
	QString HelpPath;
	QString MediaDir;

	QRect ScreenRect;
	QFontDatabase *FontDatabase;

	TBook *AddressBook;

	void FlushIni();

	uint32_t InitFido();
	void DoneFido();

	QString CharsetForMessages();

	QPixmap *FindPixmap(QString Name);

	void ChangeTextCodec(QString);
	QString toUTF(char*, QString);
	QString toUTF(char*, uint32_t = 0);
	QByteArray fromUTF(QString);

	int GetInt(QString, QString, int);
	bool GetBool(QString , QString, bool);
	QString GetStr(QString, QString, QString = QString::null);

	void GetStrList(QString, QString, QStringList*);

	QString GetDraftArea();
	QString GetArchiveArea();

	void GetPos(QWidget*);
	void GetPos(QWidget*, int, int);
	void GetSize(QWidget*, int, int);
	void GetSplitter(QString, QSplitter*, const int* = NULL);
	void GetList(QString, QListView*, const int* = NULL);
	QString GetAccel(QString, QAction*, QString);
	void GetStyleSheetItem(QString, QStyleSheetItem*, QColor, int = 0);
	void GetToolBar(QWidget*, QToolBar *Value);

	void SetInt(QString, QString, int);
	void SetBool(QString, QString, bool);
	void SetStr(QString, QString, QString);

	void SetStrList(QString, QString, QStringList*);

	void SetDraftArea(QString);
	void SetArchiveArea(QString);

	void SetPos(QWidget*);
	void SetSize(QWidget*);
	void SetSplitter(QString, QSplitter*);
	void SetList(QString, QListView*);
	void SetAccel(QString, QAction*);
	void SetStyleSheetItem(QString, QStyleSheetItem*);
	void SetToolBar(QWidget*, QToolBar *Value);

	bool IsDevelopMode();

	QString GetAddress(uint32_t);
	inline QString GetEchoTossLog()
	{
		return FidoConfig->echotosslog;
	}
	inline QString GetSysopName()
	{
		return FidoConfig->sysop;
	}

private:
	bool Inited;

	int FidoConfLock;

	TIniSettings *Settings;
	QTextCodec *TextCodec;

	void RestoreFile(QString);

	inline QString GetSection(QString Section)
	{
		return (Section.isEmpty() ? CONFIG_SECTION_MAIN : Section);
	}

	void CheckTemplates(QString);
};

#endif /* _TCONFIG_H_ */
